clients/src/main/java/org/apache/kafka/clients/admin/ListConsumerGroupsResult.java

```diff
@@ -34,70 +32,77 @@
  */
 @InterfaceStability.Evolving
 public class ListConsumerGroupsResult {
-    private final Map<Node, KafkaFutureImpl<Collection<ConsumerGroupListing>>> futuresMap;
-    private final KafkaFuture<Collection<ConsumerGroupListing>> flattenFuture;
-    private final KafkaFuture<Void> listFuture;
+    private final KafkaFutureImpl<Collection<ConsumerGroupListing>> all;
+    private final KafkaFutureImpl<Collection<ConsumerGroupListing>> valid;
+    private final KafkaFutureImpl<Collection<Throwable>> errors;
 
-    ListConsumerGroupsResult(final KafkaFuture<Void> listFuture,
-                             final KafkaFuture<Collection<ConsumerGroupListing>> flattenFuture,
-                             final Map<Node, KafkaFutureImpl<Collection<ConsumerGroupListing>>> futuresMap) {
-        this.flattenFuture = flattenFuture;
-        this.listFuture = listFuture;
-        this.futuresMap = futuresMap;
-    }
-
-    private class FutureConsumerGroupListingIterator extends AbstractIterator<KafkaFuture<ConsumerGroupListing>> {
-        private Iterator<KafkaFutureImpl<Collection<ConsumerGroupListing>>> futuresIter;
-        private Iterator<ConsumerGroupListing> innerIter;
-
-        @Override
-        protected KafkaFuture<ConsumerGroupListing> makeNext() {
-            if (futuresIter == null) {
-                try {
-                    listFuture.get();
-                } catch (Exception e) {
-                    // the list future has failed, there will be no listings to show at all
-                    return allDone();
-                }
-
-                futuresIter = futuresMap.values().iterator();
-            }
-
-            while (innerIter == null || !innerIter.hasNext()) {
-                if (futuresIter.hasNext()) {
-                    KafkaFuture<Collection<ConsumerGroupListing>> collectionFuture = futuresIter.next();
-                    try {
-                        Collection<ConsumerGroupListing> collection = collectionFuture.get();
-                        innerIter = collection.iterator();
-                    } catch (Exception e) {
-                        KafkaFutureImpl<ConsumerGroupListing> future = new KafkaFutureImpl<>();
-                        future.completeExceptionally(e);
-                        return future;
-                    }
+    ListConsumerGroupsResult(KafkaFutureImpl<Collection<Object>> future) {
+        this.all = new KafkaFutureImpl<>();
+        this.valid = new KafkaFutureImpl<>();
+        this.errors = new KafkaFutureImpl<>();
+        future.whenComplete(new KafkaFuture.BiConsumer<Collection<Object>, Throwable>() {
+            @Override
+            public void accept(Collection<Object> objects, Throwable throwable) {
+                if (throwable != null) {
```

Comment:
`throwable` should never be null, since we only call `future.complete(results);` above, in that case, we can throw an illegal state exception here.

Reply:
Good point... we always call `KafkaFuture#complete`, never `KafkaFuture#completeExceptionally`.  So this can be simplified.